### ISSUE TITLE
torch.masked_select

### ISSUE THREAD

**Issue by tianrluo** [pytorch/pytorch#31895]
## 🐛 Bug

Unexpected behavior of `masked_select`, likely due to `out` option should be contiguous but was not clarified in the doc.

## To Reproduce

```python
x, y, z = torch.ones((4,3,2)), torch.zeros((4,3), dtype=torch.bool), torch.zeros((4,2,2))
y[:, 1] = True

torch.masked_select(x, y[..., None], out=z[:, 1, :])
```

## Expected behavior

```python
>>> print(z)
tensor([[[0., 0.],  
         [1., 1.]], 
                    
        [[0., 0.],  
         [1., 1.]], 
                    
        [[0., 0.],  
         [1., 1.]], 
                    
        [[0., 0.],  
         [1., 1.]]])
```

## Observed behavior

```python
>>> print(z)
tensor([[[0., 0.],  
         [1., 1.]], 
                    
        [[1., 1.],  
         [1., 1.]], 
                    
        [[1., 1.],  
         [0., 0.]], 
                    
        [[0., 0.],  
         [0., 0.]]])
```

## Environment
PyTorch version: 1.3.1
CUDA used to build PyTorch: 10.1.243
OS: Ubuntu 18.04
Python version: 3.7

**Comment by zou3519** [comment]
I am marking this as hi-pri because it is a silent correctness issue.

**Comment by peterbell10** [comment]
In `THTensor_(maskedSelectBool)`, the output tensor is unconditionally resized to a 1d array of size `mask.sum()`:
https://github.com/pytorch/pytorch/blob/904ab092c286a750a270b4fd66ffd3a8baf39302/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L111

So, the expected behaviour is impossible since it expects the output to be a 2d tensor. However, even if output was a 1d tensor, the strides would still not be preserved by the resize.

Also note that the actual output returned is correct, but only if you read it from the `tensor` object that was passed into `masked_select`:
```python
out = z[:,1,:]
torch.masked_select(x, m[...,None], out=out)
# out = tensor([1., 1., 1., 1., 1., 1., 1., 1.])
```

The larger issue here is that `out` arguments in `torch` don't behave the same as in `numpy`. If the `out` tensor isn't the right size, `torch` functions will often resize `out` in a manner that may write anywhere in its `THStorage` object. Further, they can also allocate new storage or resize existing storage. That's true not only of legacy TH ops but also supported explicitly in `TensorIterator` for back-compat. An example with `torch.cos`:
```python
>>> a = torch.rand((2, 2))
>>> out = torch.zeros((2, 2, 2))
>>> torch.cos(a, out=out[1, 1, :])
tensor([[0.8327, 0.6649],
        [0.8733, 0.9042]])
>>> out
tensor([[[0.0000, 0.0000],
         [0.0000, 0.0000]],

        [[0.0000, 0.0000],
         [0.8327, 0.6649]]])
>>> out.numel()
8
>>> out.storage.size()
10
```

So, if `masked_select` is brought in line with other torch functions, it would mean that strides are respected only in the special case where `out.numel() == mask.sum()` and `out.ndim == 1`.

**Comment by ezyang** [comment]
> the strides would still not be preserved by the resize

Err, this is not true?

```
>>> x = torch.zeros(2, 2).t()
>>> x.stride()
(1, 2)
>>> x.resize_(2, 2)
tensor([[0., 0.],
        [0., 0.]])
>>> x.stride()
(1, 2)
```

>  The larger issue here is that out arguments in torch don't behave the same as in numpy. If the out tensor isn't the right size, torch functions will often resize out in a manner that may write anywhere in its THStorage object. 

Yes, this is a general problem with PyTorch. Arguably the master issue: https://github.com/pytorch/pytorch/issues/5272

**Comment by peterbell10** [comment]
>> the strides would still not be preserved by the resize

>Err, this is not true?

@ezyang your example only works because the tensor is already the correct size, so no resizing is actually done. Consider instead:
```python
>>> x = torch.zeros(10)[::2]                                               
>>> x.stride()                                                             
(2,)
>>> x.resize_(3)                                                           
tensor([0., 0., 0.])
>>> x.stride()                                                   
(1,)
>>> x.storage().size()
10
```

**Comment by pearu** [comment]
@peterbell10 @ezyang Here's my analysis of the issue:

First, the [doc of masked_select](https://pytorch.org/docs/stable/torch.html#torch.masked_select) does not specify any conditions to the `out` tensor parameter but from the fact that the result of `masked_select` is a 1-D tensor, I would imply that `out` is expected to be a 1-D tensor as well. 

Second, when the user specifies the `out` parameter, then 
```
masked_select(input, mask, out=out)
```
and
```
result = masked_select(input, mask)
out[:] = result
```
ought to be equivalent, in general, right?

When the `out` parameter is a non-contiguous or not a 1-D tensor, the user expects the following behavior (see the original description of this issue):
```
result = masked_select(input, mask)
result.resize_(out.shape)
out[:] = result
```
However, the current implementation in pytorch is equivalent to
```
result = masked_select(input, mask)
out.resize(result.shape)
out[:] =result
```

So, the difference in the user expectation and the current implementation can be manifested in the following choices if the user input or user output (that the `out` parameter essentially is) does not match exactly with the pytorch expectations:
1. adapt the user input to pytorch expectation,
2. adapt the pytorch result to user output.

In the case of `masked_select`, pytorch makes the choice 1. which would be indeed correct and expected for *true* inputs.
However, the `out` parameter is essentially a return value with the extension that the user gives a hint to the `masked_select` function (or in general, to any pytorch function that has optional `out` argument) where (the storage info) and how (the shape info) the results should land. So, choice 2 would be expected for `out` parameters. Currently, pytorch overrides the user given shape info leading to issues like this and at https://github.com/pytorch/pytorch/issues/5272.

Possible solutions:
1. Document the current pytorch behavior when the properties (shape, type, etc) of the `out` parameter does not match with the expected properties of a return value.
2. Do Solution 1 and raise an exception whenever the user output does not match with a pytorch return value.
3. Reconsider the pytorch current behavior where `out` parameters are treated as user inputs rather than user outputs.

While Solution 3 would be a nice feature, IMHO, and in long term would eliminate all issues alike to this one, its impact to other pytorch parts is unclear to me atm and would require additional analysis before implementation.

Thoughts?

**Comment by peterbell10** [comment]
IMO `foo(..., out=x)` should only work if `x[:] = foo(...)` would work. I believe the `out` argument is primarily intended to reduce memory allocations and resizing defeats that purpose. Perhaps an argument could be made for resizing down, since that doesn't require reallocation. However, in cases like the original issue where you're writing into a strided view of a larger tensor then any resizing is likely a bug in the user's code.

`masked_select` is particularly tricky since you don't know statically what the size of the output even is. So maybe it shouldn't have an `out` argument at all.

That said, if the current resizing model is useful to some users then how about:
- Deprecate the resizing of tensor `out` arguments
- Allow `out` to be a `Storage` object. In which case, the result is equivalent to (but without the temporary `result`):
```python
result = foo(...)
tensor_out = torch.empty(()).set_(source=out, size=result.shape)
tensor_out[:] = result
return tensor_out
```

The user has to call it as `out=tensor.storage()` which makes it clear that the operation modifies the entire storage object, not just the elements of the tensor. It also doesn't ever pretend to be strided the same as the original tensor. So, in that sense, the behaviour is more consistent.

**Comment by ezyang** [comment]
I agree with @peterbell10's point that the fact that the output size of `masked_select` is unknown is very suspicious.

> When the out parameter is a non-contiguous or not a 1-D tensor, the user expects the following behavior (see the original description of this issue):

I would modify this slightly.  When the `out` tensor is non-contiguous but the same shape, I think the user expects:

```
result = masked_select(input, mask)
out[:] = result
```

I.e. the contiguous result is scattered into the noncontiguous segments of out.

When the `out` tensor is the not the same shape but compatible with the current shape, I think they expect:

```
result = masked_select(input, mask)
out[:] = result.view_as(out)
```

When the shapes are not compatible AND out isn't a zero-size tensor I don't think the behavior is defined. And so the situation with `masked_select` is that it's very easy to end up in the UB case.

@tianrluo do you think you could help clarify why you are using `out` in this situation? Do you have a known upper bound for how large the masked selection will be?

**Comment by tianrluo** [comment]
@ezyang
It's been a while since I noticed this.

I was probably trying to repetitively in-place update the same portion of a tensor (here it's `z`) in a computation loop.
The new data for this update was mask-selected from some other tensor that also changes itself during the loop.
Mask selection returns a 1-D array tensor, so, reshape was needed before updating `z`.
I thought this selection and reshaping would provoke unwanted repetitive allocations, so, attemped using the `out` option to avoid it.

My memory here is rather vague.
(Sorry for the late reply)

**Comment by ezyang** [comment]
Thanks for letting us know. Would it be accurate then, that you could have the correct size of the output tensor if you repeatedly mask-selected every iteration of the loop (from the other tensor which is changing size)?

**Comment by tianrluo** [comment]
Right, the output tensor size was constant and known.

If I remember correctly,
the loops updates data in `x` (known constant size),
`m` (constant) alwasy select the same elements from `x`,
and put them into `z`.

I might had been over-cautious about RAM management back then...

(I am writing down only based on my vague memory...
 These attempts was not recorded in my dev history as they didn't work.)

**Comment by mruberry** [comment]
I believe this will be addressed by https://github.com/pytorch/pytorch/issues/41027.
